config_converter/config_mapper/config_mapper.py

```diff
@@ -0,0 +1,198 @@
+
+Usage: Run the exec file in the bin folder with 2 arguments - the
+path to the config file to migrate, and a directory to store the output
+master agent config file in.
+
+To do:
+1. Improve the CLI. Include a --help option. Allow optional args like
+   log-level and log-filepath (see project docs).
+2. Create a mapping for log_level if there in the fluentd config file and input
+   default ones if not given
+3. Generate stats like #fields converted, ignored, and output as schema
+4. Generate logs
+5. Fix types of numbers, lists
+6. Generate tests to check if for right logs, right stats, above changes
+"""
+
+import copy
+import sys
+import yaml
+from google.protobuf import json_format
+from config_converter.config_mapper import config_pb2
+
+# each plugin (key) maps to a dictionary of all possible fields (value)
+# each sub-dict maps fluentd field names (key) to master agent fields (value)
+# each sub-dict only contains fields that stay in the same level after mapping
+_MAIN_MAP = {
+    'in_tail': {
+        'exclude_path': 'exclude_path',
+        'path': 'path',
+        'path_key': 'path_field_name',
+        'pos_file': 'checkpoint_file',
+        'refresh_interval': 'refresh_interval',
+        'rotate_wait': 'rotate_wait'
+    }
+}
+# each plugin (key) maps to a list of all possible special fields (value)
+# these fields may not belong to the same level, not have a 1:1 mapping, etc
+# https://docs.fluentd.org/parser/multiline - shows formatN works for
+# 1 <= N <= 20
+_SPECIAL_MAP = {
+    'in_tail': [
+        'format', 'format_firstline', '@type', 'multiline_flush_interval',
+        'expression'
+    ] + [f'format{i}' for i in range(1, 21)]
+}
+# each main plugin type (key) maps to a dict of required fields (value)
+# these sub-dicts are common for all plugins of a type, hence different keys
+# each sub-dict maps fluentd field names (key) to master agent fields (value)
+# each sub-dict only contains fields that go in the upper level after mapping
+_OUTER_LEVEL_MAP = {'source': {'tag': 'name', '@type': 'type'}}
+# each plugin (key) maps to a list of all possible directives (value)
+# each list contains known and allowed sub directives
+_SUPPORTED_DIRS = {'in_tail': ['parse']}
+# list of fields we do not know how to convert
+_UNSUPPORTED_FIELDS = [
+    'emit_unmatched_lines', 'enable_stat_watcher', 'enable_watch_timer',
+    'encoding', 'from_encoding', 'ignore_repeated_permission_error',
+    'limit_recently_modified', 'open_on_every_update', 'path_timezone',
+    'pos_file_compaction_interval', 'read_from_head', 'read_lines_limit',
+    'skip_refresh_on_startup'
+]
+# plugins we know how to convert
+_SUPPORTED_PLUGINS = ['in_tail']
+# plugins we do not know how to convert
+_UNSUPPORTED_PLUGINS = ['in_forward', 'in_syslog']
+
+
+def extract_root_dirs(config_obj: config_pb2.Directive) -> dict:
+    """Checks root dirs, maps with corresponding params if supported."""
+    result = {'logs_module': dict()}
```

Comment:
Suggestion: Here just do:

    logs_module = {}

and below return:

    return {'logs_module': logs_module}

Reply:
Oh yeah that is more convenient  - will change!